accounts/abi/type.go

```diff
@@ -40,53 +38,59 @@ const (
 
 // Type is the reflection of the supported argument type
 type Type struct {
-	IsSlice   bool
-	SliceSize int
+	IsSlice, IsArray bool
+	SliceSize        int
+
+	Elem *Type
+
+	Kind reflect.Kind
+	Type reflect.Type
+	Size int
+	T    byte // Our own type checking
 
-	Kind       reflect.Kind
-	Type       reflect.Type
-	Size       int
-	T          byte   // Our own type checking
 	stringKind string // holds the unparsed string for deriving signatures
 }
 
 var (
+	// fullTypeRegex parses the abi types
+	//
+	// Types can be in the format of:
+	//
+	// 	Input  = Type [ "[" [ Number ] "]" ] Name .
+	// 	Type   = [ "u" ] "int" [ Number ] .
+	//
+	// Examples:
+	//
+	//      string     int       uint       real
+	//      string32   int8      uint8      uint[]
+	//      address    int256    uint256    real[2]
 	fullTypeRegex = regexp.MustCompile("([a-zA-Z0-9]+)(\\[([0-9]*)?\\])?")
-	typeRegex     = regexp.MustCompile("([a-zA-Z]+)([0-9]*)?")
+	// typeRegex parses the abi sub types
+	typeRegex = regexp.MustCompile("([a-zA-Z]+)([0-9]*)?")
 )
 
-// NewType returns a fully parsed Type given by the input string or an error if it  can't be parsed.
-//
-// Strings can be in the format of:
-//
-// 	Input  = Type [ "[" [ Number ] "]" ] Name .
-// 	Type   = [ "u" ] "int" [ Number ] .
-//
-// Examples:
-//
-//      string     int       uint       real
-//      string32   int8      uint8      uint[]
-//      address    int256    uint256    real[2]
+// NewType creates a new reflection type of abi type given in t.
 func NewType(t string) (typ Type, err error) {
-	// 1. full string 2. type 3. (opt.) is slice 4. (opt.) size
-	// parse the full representation of the abi-type definition; including:
-	// * full string
-	// * type
-	// 	* is slice
-	//	* slice size
 	res := fullTypeRegex.FindAllStringSubmatch(t, -1)[0]
-
 	// check if type is slice and parse type.
 	switch {
 	case res[3] != "":
 		// err is ignored. Already checked for number through the regexp
 		typ.SliceSize, _ = strconv.Atoi(res[3])
-		typ.IsSlice = true
+		typ.IsArray = true
 	case res[2] != "":
 		typ.IsSlice, typ.SliceSize = true, -1
 	case res[0] == "":
 		return Type{}, fmt.Errorf("abi: type parse error: %s", t)
 	}
+	if typ.IsArray || typ.IsSlice {
+		sliceType, err := NewType(res[1])
+		if err != nil {
+			return Type{}, err
+		}
+		typ.Elem = &sliceType
+		return typ, nil
```

Comment:
This branch prematurely returns `typ` and `stringKing` is not set (it's set at the last instruction of this method). This results in the `bind` package breaking as it doesn't have access to the original type any more to convert to a Go equivalent.

Please either add the `stringKind` assignment here too, or much better, move the assignment up to be the first statement in this method.